libcontainer/state_linux.go

```diff
@@ -0,0 +1,216 @@
+	transition(containerState) error
+	destroy() error
+	status() Status
+}
+
+// destroyedState represents a container is a stopped/destroyed state.
+type destroyedState struct {
+	c *linuxContainer
+}
+
+func (b *destroyedState) status() Status {
+	return Destroyed
+}
+
+func (b *destroyedState) transition(s containerState) error {
+	switch s.(type) {
+	case *runningState:
+		b.c.state = s
+		return nil
+	case *restoredState:
+		b.c.state = s
+		return nil
+	case *destroyedState:
+		return nil
+	}
+	return newStateError(b, s)
+}
+
+func (b *destroyedState) destroy() error {
+	if !b.c.config.Namespaces.Contains(configs.NEWPID) {
+		if err := killCgroupProcesses(b.c.cgroupManager); err != nil {
+			logrus.Warn(err)
+		}
+	}
+	err := b.c.cgroupManager.Destroy()
+	if rerr := os.RemoveAll(b.c.root); err == nil {
+		err = rerr
+	}
+	b.c.initProcess = nil
+	if herr := b.runHooks(); err == nil {
+		err = herr
+	}
+	return err
+}
+
+func (b *destroyedState) runHooks() error {
+	if b.c.config.Hooks != nil {
+		s := configs.HookState{
+			Version: b.c.config.Version,
+			ID:      b.c.id,
+			Root:    b.c.config.Rootfs,
+		}
+		for _, hook := range b.c.config.Hooks.Poststop {
+			if err := hook.Run(s); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// runningState represents a container that is currently running.
+type runningState struct {
+	base containerState
+	c    *linuxContainer
+}
+
+func (r *runningState) status() Status {
+	return Running
+}
+
+func (r *runningState) transition(s containerState) error {
+	switch s.(type) {
+	case *destroyedState:
+		running, err := r.c.isRunning()
+		if err != nil {
+			return err
+		}
+		if running {
+			return newGenericError(fmt.Errorf("container still running"), ContainerNotStopped)
+		}
+		r.c.state = s
+		return nil
+	case *pausedState:
+		r.c.state = s
+		return nil
+	case *runningState, *nullState:
+		return nil
+	}
+	return newStateError(r, s)
+}
+
+func (r *runningState) destroy() error {
+	running, err := r.c.isRunning()
+	if err != nil {
+		return err
+	}
+	if running {
+		return newGenericError(fmt.Errorf("container is not destroyed"), ContainerNotStopped)
+	}
+	return r.base.destroy()
+}
+
+// pausedState represents a container that is currently pause.  It cannot be destroyed in a
+// paused state and must transition back to running first.
+type pausedState struct {
+	c *linuxContainer
+}
+
+func (p *pausedState) status() Status {
+	return Paused
+}
+
+func (p *pausedState) transition(s containerState) error {
+	switch s.(type) {
+	case *runningState:
+		p.c.state = s
+		return nil
+	case *pausedState:
+		return nil
+	}
+	return newStateError(p, s)
+}
+
+func (p *pausedState) destroy() error {
+	return newGenericError(fmt.Errorf("container is paused"), ContainerPaused)
+}
+
+// restoredState is the same as the running state but also has accociated checkpoint
+// information that maybe need destroyed when the container is stopped and destory is called.
+type restoredState struct {
```

Comment:
I don't think so because if it was one state, as soon as you checkpointed something, when the main process stops after the checkpoint the imageDir would be deleted.